src/data/polynomial.lean

```diff
@@ -460,10 +408,107 @@ by rw [← C_1, comp_C]
 @[simp] lemma one_comp : comp (1 : polynomial R) p = 1 :=
 by rw [← C_1, C_comp]
 
+@[simp] lemma add_comp : (p + q).comp r = p.comp r + q.comp r := eval₂_add _ _
+
+end comp
+
+/-!
+We next prove that eval₂ is multiplicative
+as long as target ring is commutative
+(even if the source ring is not).
+-/
+section eval₂
+variables [comm_semiring S]
+variables (f : R → S) [is_semiring_hom f] (x : S)
+open is_semiring_hom
+
+@[simp] lemma eval₂_mul : (p * q).eval₂ f x = p.eval₂ f x * q.eval₂ f x :=
+begin
+  dunfold eval₂,
+  rw [mul_def, finsupp.sum_mul _ p], simp only [finsupp.mul_sum _ q], rw [sum_sum_index],
+  { apply sum_congr rfl, assume i hi, dsimp only, rw [sum_sum_index],
+    { apply sum_congr rfl, assume j hj, dsimp only,
+      rw [sum_single_index, map_mul f, pow_add],
+      { simp only [mul_assoc, mul_left_comm] },
+      { rw [map_zero f, zero_mul] } },
+    { intro, rw [map_zero f, zero_mul] },
+    { intros, rw [map_add f, add_mul] } },
+  { intro, rw [map_zero f, zero_mul] },
+  { intros, rw [map_add f, add_mul] }
+end
+
+lemma eval₂_smul (g : R →+* S) (p : polynomial R) (x : S) {s : R} :
```

Comment:
This one is also true without the commutativity assumption, isn't it?
(If so, I'm fine with just leaving a TODO.)

Reply:
Generalised!